R/insert_into_returning.R

```diff
@@ -8,88 +8,181 @@
 #' @param key Keys to check for already existing in the database. If trying to
 #' add a new row with unique key constraints use this to check for existing
 #' entries with same information.
-#' @param id The field to return, defaults to using "id"
+#' @param return The field to return, defaults to using "id"
 #'
 #' @return List of returned ids from inserted values.
 #'
 #' @keywords internal
-insert_into_returning <- function(con, table, d, key = NULL, id = NULL) {
+insert_into_returning <- function(con, table, d, key = NULL, return = NULL) {
   UseMethod("insert_into_returning")
 }
 
 insert_into_returning.SQLiteConnection <- function(con, table, d, key = NULL,
-                                               id = NULL) {
-  id <- id %||% (if (length(key) == 1L) key else "id")
-  stopifnot(length(id) == 1L)
+                                               return = NULL) {
+  return <- return %||% (if (length(key) >= 1L) key else "id")
   insert1 <- function(i) {
     x <- as.list(d[i, , drop = FALSE])
     x <- x[!vlapply(x, is.na)]
-    sql <- c(sprintf("INSERT INTO %s", table),
-             sprintf("  (%s)", paste(names(x), collapse = ", ")),
-             "VALUES",
-             sprintf("  (%s)", paste0("$", seq_along(x), collapse = ", "))
-            )
-    sql <- paste(sql, collapse = "\n")
-    if (is.null(key)) {
-      DBI::dbExecute(con, sql, unname(x))
-      ## TODO: What happens here if we insert but the table has no rowid?
-      rowid_last_entry <- DBI::dbGetQuery(con, "SELECT last_insert_rowid()")[1, 1]
-      DBI::dbGetQuery(con, sprintf("SELECT %s FROM %s WHERE rowid = $2", id, table), rowid_last_entry)[[1]]
+    if (length((names(x))) > 0) {
```

Comment:
`length((names(x)))` should be `ncol(x)` I think, and that `as.list` can go away (otherwise just `length(x)`)